<a href="https://colab.research.google.com/github/lukaszplust/data-science-learn/blob/main/141_150_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Pandas

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 141](#1)
* [Ćwiczenie 142](#2)
* [Ćwiczenie 143](#3)
* [Ćwiczenie 144](#4)
* [Ćwiczenie 145](#5)
* [Ćwiczenie 146](#6)
* [Ćwiczenie 147](#7)
* [Ćwiczenie 148](#8)
* [Ćwiczenie 149](#9)
* [Ćwiczenie 150](#10)

### <a name='0'></a>Import biblioteki

In [1]:
import numpy as np
import pandas as pd

np.random.seed(42)
np.__version__

'1.21.6'

### <a name='1'></a> Ćwiczenie 141
Podane są dwa obiekty typu _Series_:

```
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))
```
Połącz te dwa obiekty w jeden obiekt typu _DataFrame_ (jako dwie kolumny) i przypisz do zmiennej _df_. Nadaj nazwy kolumn odpowiednio _col1_ oraz _col2_.


In [12]:
s1 = pd.Series(np.random.rand(20))
s2 = pd.Series(np.random.randn(20))


df = pd.concat([s1,s2],axis=1)
df.columns = ['col1','col2']
df

col1      col2
0   0.760785  0.915402
1   0.561277  0.328751
2   0.770967 -0.529760
3   0.493796  0.513267
4   0.522733  0.097078
5   0.427541  0.968645
6   0.025419 -0.702053
7   0.107891 -0.327662
8   0.031429 -0.392108
9   0.636410 -1.463515
10  0.314356  0.296120
11  0.508571  0.261055
12  0.907566  0.005113
13  0.249292 -0.234587
14  0.410383 -1.415371
15  0.755551 -0.420645
16  0.228798 -0.342715
17  0.076980 -0.802277
18  0.289751 -0.161286
19  0.161221  0.404051

### <a name='2'></a> Ćwiczenie 142
Wytnij z obiektu _df_ wiersze, dla których wartość w kolumnie _col2_ jest pomiędzy 0.0 a 1.0 (włącznie).

In [14]:
df[df['col2'].between(0.0,1.0)]

col1      col2  col3
0   0.760785  0.915402     1
1   0.561277  0.328751     1
3   0.493796  0.513267     1
4   0.522733  0.097078     1
5   0.427541  0.968645     1
10  0.314356  0.296120     1
11  0.508571  0.261055     1
12  0.907566  0.005113     1
19  0.161221  0.404051     1

### <a name='3'></a> Ćwiczenie 143
Przypisz nową kolumnę o nazwie _col3_, która przyjmie wartość 1 w momencie gdy w kolumnie _col2_ jest wartość nieujemna i przeciwnie -1.

In [18]:
df['col3']=df['col2'].map(lambda x: 1 if x >= 0 else -1) # BARDZO PRZYDATNE WRÓCIĆ DO TEGO
df

col1      col2  col3      col4
0   0.760785  0.915402     1  0.915402
1   0.561277  0.328751     1  0.328751
2   0.770967 -0.529760    -1 -0.529760
3   0.493796  0.513267     1  0.513267
4   0.522733  0.097078     1  0.097078
5   0.427541  0.968645     1  0.968645
6   0.025419 -0.702053    -1 -0.702053
7   0.107891 -0.327662    -1 -0.327662
8   0.031429 -0.392108    -1 -0.392108
9   0.636410 -1.463515    -1 -1.000000
10  0.314356  0.296120     1  0.296120
11  0.508571  0.261055     1  0.261055
12  0.907566  0.005113     1  0.005113
13  0.249292 -0.234587    -1 -0.234587
14  0.410383 -1.415371    -1 -1.000000
15  0.755551 -0.420645    -1 -0.420645
16  0.228798 -0.342715    -1 -0.342715
17  0.076980 -0.802277    -1 -0.802277
18  0.289751 -0.161286    -1 -0.161286
19  0.161221  0.404051     1  0.404051

### <a name='4'></a> Ćwiczenie 144
Przypisz nową kolumnę o nazwie _col4_, która przytnie wartości z kolumny _col2_ do przedziału $[-1.0, 1.0]$. 
Innymi słowy, dla wartości poniżej -1.0 należy wstawić wartość -1.0, dla wartości powyżej 1.0 wartość 1.0.


In [17]:
df['col4'] = df['col2'].clip(-1.0, 1.0) # CLIP RÓWNIEŻ PRZYDATNE
df

col1      col2  col3      col4
0   0.760785  0.915402     1  0.915402
1   0.561277  0.328751     1  0.328751
2   0.770967 -0.529760    -1 -0.529760
3   0.493796  0.513267     1  0.513267
4   0.522733  0.097078     1  0.097078
5   0.427541  0.968645     1  0.968645
6   0.025419 -0.702053    -1 -0.702053
7   0.107891 -0.327662    -1 -0.327662
8   0.031429 -0.392108    -1 -0.392108
9   0.636410 -1.463515    -1 -1.000000
10  0.314356  0.296120     1  0.296120
11  0.508571  0.261055     1  0.261055
12  0.907566  0.005113     1  0.005113
13  0.249292 -0.234587    -1 -0.234587
14  0.410383 -1.415371    -1 -1.000000
15  0.755551 -0.420645    -1 -0.420645
16  0.228798 -0.342715    -1 -0.342715
17  0.076980 -0.802277    -1 -0.802277
18  0.289751 -0.161286    -1 -0.161286
19  0.161221  0.404051     1  0.404051

### <a name='5'></a> Ćwiczenie 145
Znajdź 5 największych wartości dla kolumny _col2_.

In [19]:
df['col2'].nlargest(5)

5     0.968645
0     0.915402
3     0.513267
19    0.404051
1     0.328751
Name: col2, dtype: float64

Znajdź 5 najmniejszych wartości dla kolumny _col2_.

In [20]:
df['col2'].nsmallest(5)

9    -1.463515
14   -1.415371
17   -0.802277
6    -0.702053
2    -0.529760
Name: col2, dtype: float64

### <a name='6'></a> Ćwiczenie 146
Wyznacz skumulowaną sumę dla każdej kolumny.

__Wskazówka:__ Użyj metody _pd.DataFrame.cumsum()_.

In [23]:
df.cumsum() # SKUMULOWANA SUMA WARTO WRÓCIĆ , KAŻDY WIERSZ JEST DODAWANY DO POPRZEDNIEGO ITP...!!

col1      col2  col3      col4
0   0.760785  0.915402     1  0.915402
1   1.322062  1.244153     2  1.244153
2   2.093029  0.714393     1  0.714393
3   2.586825  1.227660     2  1.227660
4   3.109558  1.324738     3  1.324738
5   3.537099  2.293383     4  2.293383
6   3.562518  1.591330     3  1.591330
7   3.670409  1.263668     2  1.263668
8   3.701839  0.871560     1  0.871560
9   4.338249 -0.591955     0 -0.128440
10  4.652605 -0.295835     1  0.167680
11  5.161176 -0.034780     2  0.428735
12  6.068742 -0.029666     3  0.433849
13  6.318034 -0.264253     2  0.199261
14  6.728417 -1.679624     1 -0.800739
15  7.483968 -2.100270     0 -1.221384
16  7.712767 -2.442984    -1 -1.564098
17  7.789747 -3.245261    -2 -2.366376
18  8.079498 -3.406547    -3 -2.527661
19  8.240719 -3.002496    -2 -2.123610

### <a name='7'></a> Ćwiczenie 147
Wyznacz medianę dla zmiennej _col2_ (inaczej kwantyl rzędu 0.5).

In [24]:
df['col2'].median()

-0.19793642252057803

### <a name='8'></a> Ćwiczenie 148
Wytnij wiersze, dla których zmienna _col2_ przyjmuje wartości większe od 0.0.

In [26]:
df[df['col2'] > 0.0]

col1      col2  col3      col4
0   0.760785  0.915402     1  0.915402
1   0.561277  0.328751     1  0.328751
3   0.493796  0.513267     1  0.513267
4   0.522733  0.097078     1  0.097078
5   0.427541  0.968645     1  0.968645
10  0.314356  0.296120     1  0.296120
11  0.508571  0.261055     1  0.261055
12  0.907566  0.005113     1  0.005113
19  0.161221  0.404051     1  0.404051

### <a name='9'></a> Ćwiczenie 149
Wytnij 5 pierwszych wierszy obiektu _df_ i przekształć je do słownika.

In [29]:
df.head(5).to_dict()

{'col1': {0: 0.7607850486168974,
  1: 0.5612771975694962,
  2: 0.770967179954561,
  3: 0.49379559636439074,
  4: 0.5227328293819941},
 'col2': {0: 0.9154021177020741,
  1: 0.32875110965968446,
  2: -0.5297602037670388,
  3: 0.5132674331133561,
  4: 0.09707754934804039},
 'col3': {0: 1, 1: 1, 2: -1, 3: 1, 4: 1},
 'col4': {0: 0.9154021177020741,
  1: 0.32875110965968446,
  2: -0.5297602037670388,
  3: 0.5132674331133561,
  4: 0.09707754934804039}}

### <a name='10'></a> Ćwiczenie 150
Wytnij 5 pierwszych wierszy obiektu _df_, przekształć je do formatu Markdown i przypisz do zmiennej _df_markdown_.

In [30]:
df_markdown = df.head(5).to_markdown()

Wydrukuj do konsoli zawartość zmiennej _df_markdown_.

In [33]:
print(df_markdown) # tu sobie zapamiętać- kopuje wynik co mi wyjdzie pod tym,klikam 2 razy na następną komórkę i wklejam i mam interaktywną :DD 

|    |     col1 |       col2 |   col3 |       col4 |
|---:|---------:|-----------:|-------:|-----------:|
|  0 | 0.760785 |  0.915402  |      1 |  0.915402  |
|  1 | 0.561277 |  0.328751  |      1 |  0.328751  |
|  2 | 0.770967 | -0.52976   |     -1 | -0.52976   |
|  3 | 0.493796 |  0.513267  |      1 |  0.513267  |
|  4 | 0.522733 |  0.0970775 |      1 |  0.0970775 |


|    |     col1 |       col2 |   col3 |       col4 |
|---:|---------:|-----------:|-------:|-----------:|
|  0 | 0.760785 |  0.915402  |      1 |  0.915402  |
|  1 | 0.561277 |  0.328751  |      1 |  0.328751  |
|  2 | 0.770967 | -0.52976   |     -1 | -0.52976   |
|  3 | 0.493796 |  0.513267  |      1 |  0.513267  |
|  4 | 0.522733 |  0.0970775 |      1 |  0.0970775 |Skopiuj wynik uruchomienia powyższej komórki i wklej poniżej:

\# tutaj wpisz rozwiązanie
